### DISCUSSION TITLE
Compatibilty guarantees for Profile-Guided Optimization (PGO) profiles generated by GraalVM

### DISCUSSION THREAD

**Discussion by zamazan4ik** [oracle/graal#7892] [Q&A]
Hi!

We are evaluating using Profile-Guided Optimization (PGO) with GraalVM. Right now the GraalVM compiler stores the profile as a file, that later can be reused. However, I cannot find the information about the backward/forward compatibility guarantees of the PGO profile file format. For us, it's important to know because right now we cannot understand whether will it be possible to use PGO profiles from the older compiler with a newer GraalVM version or not (and vice versa).

In LLVM PGO infrastructure we already [met](https://github.com/llvm/llvm-project/issues/52683) some problems with that - the Clang/LLVM documentation also [hasn't](https://discourse.llvm.org/t/profile-guided-optimization-pgo-related-questions-and-suggestions/75232/2) this information yet. We want to avoid such problems with GraalVM in the future.

Thank you in advance.

**Comment by alina-yur** [comment]
Hi. I don't think there should be any incompatibilities between the profiles from different GraalVM versions, but maybe @axel22 can confirm.

**Reply by zamazan4ik** [reply]
By the way, the question is not only about compatibilities from the "Can newer GraalVM read older PGO profiles from an older GraalVM version" but also about matching PGO profiles on the source level.

We already met in LLVM and GCC that PGO profiles with different compiler optimization options can be incompatible due to huge PGO profile reliance on the inlining decision. Even small changes in the inliner decision (like changing the inliner logic or tweaking the inliner thresholds between the compiler releases) can affect PGO profiles' reusability significantly. A similar question in Go compiler: https://github.com/golang/go/issues/64394#issuecomment-1828871450

What is the current situation with this in GraalVM?
